File: packages/react-native-nitro-lunar-range-picker/android/src/main/java/com/nitrolunarrangepicker/LunarCalendarHelper.kt
package com.nitrolunarrangepicker

import java.util.Calendar
import java.util.Locale

data class LunarDateResult(
    val day: Int,
    val month: Int,
    val year: Int,
    val isLeap: Boolean,
    val lunarDayName: String,
    val isSpecialDay: Boolean
)

object LunarCalendarHelper {
    fun convertSolarToLunar(calendar: Calendar, language: String = "vi"): LunarDateResult {
        return try {
            val uCalendar = android.icu.util.ChineseCalendar()
            uCalendar.timeInMillis = calendar.timeInMillis

            val day = uCalendar.get(android.icu.util.ChineseCalendar.DAY_OF_MONTH)
            val month = uCalendar.get(android.icu.util.ChineseCalendar.MONTH) + 1
            val year = uCalendar.get(android.icu.util.ChineseCalendar.YEAR)
            val isLeap = uCalendar.get(android.icu.util.ChineseCalendar.IS_LEAP_MONTH) == 1

            val lunarDayName = String.format(Locale.getDefault(), "%02d/%02d", day, month)
            val isSpecialDay = (day == 1 || day == 15)

            LunarDateResult(day, month, year, isLeap, lunarDayName, isSpecialDay)
        } catch (e: Exception) {
            val day = calendar.get(Calendar.DAY_OF_MONTH)
            val month = calendar.get(Calendar.MONTH) + 1
            val year = calendar.get(Calendar.YEAR)
            LunarDateResult(
                day = day,
                month = month,
                year = year,
                isLeap = false,
                lunarDayName = String.format(Locale.getDefault(), "%02d/%02d", day, month),
                isSpecialDay = (day == 1 || day == 15)
            )
        }
    }
}

File: packages/react-native-nitro-lunar-range-picker/ios/LunarCalendarHelper.swift
import Foundation

public struct LunarDateResult {
    public let day: Int
    public let month: Int
    public let year: Int
    public let isLeap: Bool
    public let lunarDayName: String
    public let isSpecialDay: Bool
}

public class LunarCalendarHelper {
    /// Chuyển đổi từ NSDate sang Âm Lịch (Sử dụng Calendar Chinese chuẩn của hệ thống)
    public static func convertSolarToLunar(date: Date, language: String = "vi") -> LunarDateResult {
        let chineseCalendar = Calendar(identifier: .chinese)
        let day: Int
        let month: Int
        let year: Int
        var isLeap = false

        if #available(iOS 17.0, *) {
            let components = chineseCalendar.dateComponents([.year, .month, .day, .isLeapMonth], from: date)
            day = components.day ?? 1
            month = components.month ?? 1
            year = components.year ?? 1
            isLeap = components.isLeapMonth ?? false
        } else {
            let components = chineseCalendar.dateComponents([.year, .month, .day], from: date)
            day = components.day ?? 1
            month = components.month ?? 1
            year = components.year ?? 1
        }
        
        // Hiển thị ngày âm lịch dạng DD/MM (ví dụ 01/08, 15/08)
        let dayName = String(format: "%02d/%02d", day, month)
        let isSpecial = (day == 1 || day == 15)
        
        return LunarDateResult(
            day: day,
            month: month,
            year: year,
            isLeap: isLeap,
            lunarDayName: dayName,
            isSpecialDay: isSpecial
        )
    }
}